homeassistant/components/shelly/number.py

```diff
@@ -0,0 +1,142 @@
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.entity_registry import RegistryEntry
+
+from .const import AIOSHELLY_DEVICE_TIMEOUT_SEC, CONF_SLEEP_PERIOD
+from .entity import (
+    BlockEntityDescription,
+    ShellySleepingBlockAttributeEntity,
+    async_setup_entry_attribute_entities,
+)
+from .utils import get_device_entry_gen
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+@dataclass
+class BlockNumberDescription(BlockEntityDescription, NumberEntityDescription):
+    """Class to describe a BLOCK sensor."""
+
+    mode: NumberMode = NumberMode("slider")
+    rest_path: str = ""
+    rest_arg: str = ""
+
+
+NUMBERS: Final = {
+    ("device", "valvePos"): BlockNumberDescription(
+        key="device|valvepos",
+        icon="mdi:pipe-valve",
+        name="Valve Position",
+        unit_of_measurement=PERCENTAGE,
+        available=lambda block: cast(int, block.valveError) != 1,
+        entity_category=EntityCategory.CONFIG,
+        min_value=0,
+        max_value=100,
+        step=1,
+        mode=NumberMode("slider"),
+        rest_path="thermostat/0",
+        rest_arg="pos",
+    ),
+}
+
+
+def _build_block_description(entry: RegistryEntry) -> BlockNumberDescription:
+    """Build description when restoring block attribute entities."""
+    assert entry.capabilities
+    return BlockNumberDescription(
+        key="",
+        name="",
+        icon=entry.original_icon,
+        unit_of_measurement=entry.unit_of_measurement,
+        device_class=entry.original_device_class,
+        min_value=cast(float, entry.capabilities.get("min")),
+        max_value=cast(float, entry.capabilities.get("max")),
+        step=cast(float, entry.capabilities.get("step")),
+        mode=cast(NumberMode, entry.capabilities.get("mode")),
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up numbers for device."""
+    if get_device_entry_gen(config_entry) == 2:
+        return
+
+    if config_entry.data[CONF_SLEEP_PERIOD]:
+        await async_setup_entry_attribute_entities(
+            hass,
+            config_entry,
+            async_add_entities,
+            NUMBERS,
+            BlockSleepingNumber,
+            _build_block_description,
+        )
+
+
+class BlockSleepingNumber(ShellySleepingBlockAttributeEntity, NumberEntity):
+    """Represent a block sleeping number."""
+
+    entity_description: BlockNumberDescription
+
+    @property
+    def value(self) -> float:
+        """Return value of number."""
+        if self.block is not None:
+            return cast(float, self.attribute_value)
+
+        return cast(float, self.last_state)
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return unit of number."""
+        return self.entity_description.unit_of_measurement
+
```

Comment:
Is the unit_of_measurement still needed? I'd have thought that would be handled by the entity_description automatically.

Reply:
I don't find it as part of `ShellySleepingBlockAttributeEntity` or `NumberEntity`

Reply:
It's part of the base Entity:
https://github.com/home-assistant/core/blob/0cabf3e5770d69ccb6e246bca678c477f31f7a3d/homeassistant/helpers/entity.py#L394-L401

Reply:
It is part of the base entity: https://github.com/home-assistant/core/blob/0cabf3e5770d69ccb6e246bca678c477f31f7a3d/homeassistant/helpers/entity.py#L308

Reply:
Suggest to just remove it:
```suggestion
```